### PR TITLE
WIP: Add FsCheck project in preparation of Hackathon

### PR DESCRIPTION
**Pull request by forki**
As @Andrea described in #360 we want to create FsCheck tests for FSharp.Core.dll during the [F# eXchange Hack Day](https://skillsmatter.com/conferences/7123-f-exchange-hack-day).

This PR will prepare the solution for the participants to get started easily.

* [ ] SoundCheck works
* [ ] Create test script which only calls the property tests
* [ ] Create readme entry for the hackathon
* [ ] Rebase/Squash

### CODE REVIEW

**Comment by forki** [comment]
------ Test started: Assembly: FSharp.Core.PropertyTests.dll ------
    
    Test 'FSharp.Core.PropertyTests.SoundCheck+SoundCheck.ShouldFail' failed: System.IO.FileLoadException : Die Datei oder Assembly "FsCheck, Version=1.0.4.0, Culture=neutral, PublicKeyToken=null" oder eine Abhängigkeit davon wurde nicht gefunden. Eine Assembly mit starkem Namen ist erforderlich. (Ausnahme von HRESULT: 0x80131044)
    	at FSharp.Core.PropertyTests.SoundCheck.SoundCheck.ShouldFail()
    
    0 passed, 1 failed, 0 skipped, took 0,22 seconds (NUnit 2.6.2).


I copied the FSHarp.Core.Unittest project and it seems the code wants a FsCheck.dll with a strong name. Can we work around this?

/cc @kurtschelfthout @latkin @KevinRansom

File: src/fsharp.sln [hunk 1]
```diff
@@ -33,54 +33,257 @@ Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "FsiAnyCPU", "fsharp\fsiAnyC
...
+		Release|Mixed Platforms = Release|Mixed Platforms
+		VSDebug|Any CPU = VSDebug|Any CPU
+		VSDebug|Mixed Platforms = VSDebug|Mixed Platforms
+		VSRelease|Any CPU = VSRelease|Any CPU
+		VSRelease|Mixed Platforms = VSRelease|Mixed Platforms
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
 		{C94C257C-3C0A-4858-B5D8-D746498D1F08}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
 		{C94C257C-3C0A-4858-B5D8-D746498D1F08}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{C94C257C-3C0A-4858-B5D8-D746498D1F08}.Debug|Mixed Platforms.ActiveCfg = Debug|Any CPU
```
Comment (forki) [inline src/fsharp.sln]: I will never understand what VS is doing with the sln file...

**Comment by kurtschelfthout** [comment]
Guess we need a strongly signed build of FsCheck. Are these tests also using NUnit or xUnit?

**Comment by forki** [comment]
Only nunit

**Comment by dsyme** [comment]
Remove the strong name from FSharp.Core.Unjttests itself? It shouldn't need to be strong named, it's just a test DLL

**Comment by forki** [comment]
That's why I asked. Why are we strong naming it? I'm not sure what all the
flags are doing and if they are needed.

**Comment by forki** [comment]
> Remove the strong name from FSharp.Core.Unjttests itself?

any tips where to look?

**Comment by forki** [comment]
I got rid of the strong names, but now it seems I get an FSharp.Core incompatibility issue.
Any ideas?

    ------ Test started: Assembly: FSharp.Core.PropertyTests.dll ------
    
    Test 'FSharp.Core.PropertyTests.SoundCheck+SoundCheck.ShouldFail' failed: System.InvalidCastException : Das Objekt des Typs "Arrow@207-2[Microsoft.FSharp.Collections.FSharpList`1[System.Int32],System.Boolean]" kann nicht in Typ "Testable`1[Microsoft.FSharp.Core.FSharpFunc`2[Microsoft.FSharp.Collections.FSharpList`1[System.Int32],System.Boolean]]" umgewandelt werden.
    	at Microsoft.FSharp.Core.LanguagePrimitives.IntrinsicFunctions.UnboxGeneric[T](Object source)
    	at FsCheck.Runner.check[a](Config config, a p)
    	at FsCheck.Check.QuickThrowOnFailure[Testable](Testable property)
    	D:\code\fsharp\src\fsharp\FSharp.Core.PropertyTests\SoundCheck.fs(17,0): at FSharp.Core.PropertyTests.SoundCheck.SoundCheck.ShouldFail()
    
    0 passed, 1 failed, 0 skipped, took 0,33 seconds (NUnit 2.6.2).

**Comment by kurtschelfthout** [comment]
That's very likely an issue where FsCheck references a different version of FSharp.Core than the one you're testing with. Usually assembly redirects work, but given this is testing FSharp.Core itself I'm frankly a bit confused how that all should work. Are the other test projects dependent on something that is itself dependent on FSharp.Core?

**Comment by forki** [comment]
> Are the other test projects dependent on something that is itself dependent on FSharp.Core?

nope.

> Usually assembly redirects work

how would you create these for a dll?

**Comment by kurtschelfthout** [comment]
You don't, for NUnit you have to create them for the runner or whatever the exe is that executes the tests. If I remember correctly only xunit is smart enough to load redirects specified with the dll in it's test assembly domain.

**Comment by kurtschelfthout** [comment]
You probably realize this but to be clear: you should try to redirect to the built version, not an actual released version...otherwise the assembly under test will be redirected to be the release version too...

**Comment by forki** [comment]
quick question: could we ILMerge the Fsharp.core into FsCheck?

**Comment by kurtschelfthout** [comment]
What is this magic you speak of :)  haven't used ILMerge before.

**Comment by forki** [comment]
look at https://github.com/fsprojects/Paket/blob/master/build.fsx#L158 - we merge FSharp.Core into paket.exe and all these weird issues go away. Would you accept such a pull request?

**Comment by kurtschelfthout** [comment]
I don't think that's actually a solution here (or even for FsCheck generally). I understand that for an exe, this is the way to go. But here, we want there to be only one version of FSharp.Core in play, and that version should be the one that is just built. Whether FsCheck references or merges another one, I don't think will make any difference. This is more of a bootstrap problem...we are testing something that is itself being used to write the tests....

**Comment by enricosada** [comment]
@forki you need to add an assemblyname.dll.config for redirect binding of assembly. see https://github.com/Microsoft/visualfsharp/blob/fsharp4/vsintegration/src/unittests/Unittests.dll.config

**Comment by forki** [comment]
@enricosada I'm not sure I understand. could you please help me out with a commit (and pull request to my branch) or a bit more details?

**Comment by forki** [comment]
like  2e442b0?

**Comment by forki** [comment]
HA!

    ------ Test started: Assembly: FSharp.Core.PropertyTests.dll ------
    
    Ok, passed 100 tests.
    
   1 passed, 0 failed, 0 skipped, took 0,37 seconds (NUnit 2.6.2).

owl

**Comment by enricosada** [comment]
man, you are too fast. god job!

**Comment by kurtschelfthout** [comment]
Appropriate typo :)
:+1:

**Comment by forki** [comment]
Now what's the easiest way to get the PropertyAttribute working. It doesn't seem to pickup my tests when I convert the sound check.

**Comment by kurtschelfthout** [comment]
Have a look at FsCheck.NUnitExamples. Needs to reference FsCheck.NUnint, FsCheck.NUnit.Addin, and needs to add the FsCheckAddin.fs file. Hopefully that should do the trick.

https://github.com/fsharp/FsCheck/tree/master/examples/FsCheck.NUnit.Examples

**Comment by forki** [comment]
strange. your paket.references says: reference NUnit, but in the fsproj you do: https://github.com/fsharp/FsCheck/blob/master/examples/FsCheck.NUnit.Examples/FsCheck.NUnit.Examples.fsproj#L69

**Comment by forki** [comment]
also it doesn't have a version:

image

**Comment by kurtschelfthout** [comment]
I _think_ this would all work out OK if that project would use the actual released FsCheck and FsCheck.NUnit, but it only references NUnit via paket. In other words, the other references are added manually so it can reference the project files, not binaries.

**Comment by forki** [comment]
Yep, but you should at least add nunit runners to references file.

**Comment by kurtschelfthout** [comment]
I see, yes, agreed.

**Comment by mavnn** [comment]
If it helps, I never used the NUnit plugins (they weren't there when I first starting with FsCheck!). Since then I've always just used a test with Check.All and not bothered with the extra dependencies: https://github.com/mavnn/MAG/blob/master/MAG.Properties/Serialization.fs#L31

**Comment by forki** [comment]
yep. since I don't get the other thing to work. this will be the way for the hackday.

**Comment by latkin** [comment]
@forki @andrea does this PR need to stay open for the event or can it be closed out in lieu of a consolidated PR coming later?

**Comment by forki** [comment]
I only needed it to see if I can make the CI happy.

**Pull request closed without merging**